collections-und-generics/README.md

```diff
@@ -1,3 +1,451 @@
 # Collections und Generics
+## Inhalt
+* Generische Methoden und Klassen
+* Constraints in C#
+* Was sind Collections
+* Generische Interfaces
+* Einsatz von Generics in Collections als Datenstruktur
+* Quellen  
 
-[Zur Übersicht](../README.md)
\ No newline at end of file
+## Generische Methoden und Klassen
+
+### Generics Bedeutung
+Generisch bedeutet im Allgemeinen gültigen Sinne.
+
+#### Motivation:
+Man versucht folgende vier "Hauptziele" zu erreichen:
+ * Die Typsicherheit zu erhöhen
+ * Type-Casts nicht durchführen zu müssen
+ * Box'ing und Unbox'ing nicht durchführen zu müssen
+ * C# bzw. andere OO-Programiersprachen haben viele 'Templates'
+
+ #### Vorteile gegenüber der verwendung der allgemein Objekt-Klasse:
+ * Es entsteht ein klarer Performancegewinn
+ * Der Sourcecode ist wiederverwendbar
+
+### Generische Klassen 
+'T'oder 'E' werden im Allgemein als die generische Typevariablen verwendent.
+Bei der folgende Zeile wird gezeigt, wie eine generische Klasse aufgebaut ist:
+~~~cs
+access_modifer class ident <T,K> // T, K generic datatType
+~~~
+Generische Typen können in internen Fields, Properites und Methoden einer Klasse verwendet werden.
+~~~cs
+public  class Kunde<T>  //TKundenInfo anstelle von T
+{                       //wäre geeigneter 
+    private static List<T> _kundenListe;
+    private T _kundenInfo;
+    public T KundenInfo{ get; set; };
+    public int VerlgleicheKunden(T KundenInfo);
+}
+~~~
+Bei der Verwendung von generischen Typen ist als es ob man irgend einen anderen nichtgenerischen Typ verwendet.
+Die Typ-Parameter müssen lediglich bei der Instanzierung gesetzt werden.
+~~~cs
+LinkedList<string> myList = new LinkedList<string>();
+~~~
+## Constraints in C#
+Das "Keyword", ***where*** bietet Beschränkungen für einen Typ-Parameter.
+Es können auch ein oder mehrere Basisklassen oder ein Interfaces verwendet werden als Einschränkung.
+### Beispiel:
+~~~cs
+public class MyGenericClass<T> where T : IComparable<T> { }
+~~~
+Hier muss die die Klasse MyGenericClass als Einschränkung das IComarableInterface implementieren. Also ein 'Vertrag' schreibt vor, dass der Typ T 'vergleichbar' sein muss gemäß dem Interface.
+### Einen Default Wert verwenden
+Das "Keyword" ***default*** 
+Wenn T ein Referenztyp ist dann ist defaut(T) *null*;
+Für Werttypen wie zum Beispiel int, float,  hier werden alle Bits werden auf *0* gesetzt.
+
+## Was sind Collections?
+Collections sind aufzählbare Datenstrukturen die ausgewertet werden können indem man Indexe oder Schlüssel (Keys) verwendet. Datenstrukturen welche nahe in Relation stehen können effizienter gehandhabt werden als jene in einer Collection.
```

Comment:
Diesen Satz verstehe ich nicht:
"Datenstrukturen welche nahe in Relation stehen können effizienter gehandhabt werden als jene in einer Collection."